using System.Net;

namespace ContactManagement.Middleware
{
	public class CustomExceptionMiddleware
	{
			private readonly RequestDelegate _next;
			private readonly ILogger<CustomExceptionMiddleware> _logger;

			public CustomExceptionMiddleware(RequestDelegate next, ILogger<CustomExceptionMiddleware> logger)
			{
				_next = next;
				_logger = logger;
			}

			public async Task InvokeAsync(HttpContext httpContext)
			{
				try
				{
					await _next(httpContext);
				}
				catch (Exception ex)
				{
					_logger.LogError($"Something went wrong: {ex}");
					await HandleExceptionAsync(httpContext, ex);
				}
			}

			private Task HandleExceptionAsync(HttpContext context, Exception exception)
			{
				context.Response.ContentType = "application/json";
				context.Response.StatusCode = (int)HttpStatusCode.InternalServerError;

				return context.Response.WriteAsync(new
				{
					StatusCode = context.Response.StatusCode,
					Message = "Internal Server Error from the custom middleware."
				}.ToString());
			}
		}
	}
